transient/image.py

```diff
@@ -49,37 +49,40 @@ def retrieve_image(
         # Do partial downloads into the working directory in the backend
         dest_name = os.path.basename(destination)
         temp_destination = os.path.join(store.working, dest_name)
-        fd = self.__lock_backend_destination(temp_destination)
 
-        # We now hold the lock. Either another process started the retrieval
-        # and died (or never started at all) or they completed. If the final file exists,
-        # the must have completed successfully so just return.
-        if os.path.exists(destination):
-            logging.info("Retrieval completed by another processes. Skipping.")
-            os.close(fd)
-            return None
+        with utils.cleanup_on_error(
+            temp_destination,
+            "wb+",
+            opener=self.__lock_backend_destination,
+            rename=destination,
+        ) as temp_file:
+            # We now hold the lock. Either another process started the retrieval
+            # and died (or never started at all) or they completed. If the final file exists,
+            # they must have completed successfully so just return.
+            if os.path.exists(destination):
+                logging.info("Retrieval completed by another processes. Skipping.")
+                os.unlink(temp_destination)
```

Comment:
Hmm, will we still attempt to rename `temp_destination` to `destination` in the context manager in this case? That will cause an exception, right?

Reply:
I think that's right. Made a fix for that. This will be tough to test in behave, is there a unit testing setup for transient?

Reply:
Not currently. I'd be ok depending on https://nose.readthedocs.io/en/latest/ to do that

Reply:
Hmm, nose has been abandoned for several years now. I've had really good experiences with [pytest](https://docs.pytest.org/en/6.2.x/).

Reply:
Oh dang, apparently it has. Pytest works as well.